### DISCUSSION TITLE
Release 0.7.x (Security & scoped package)

### DISCUSSION THREAD

**Discussion by karfau** [xmldom/xmldom#270] [General]
**This package is now published to npm as `@xmldom/xmldom`!**

### Fixed

- dom: prevent iteration over deleted items [`#514`](https://github.com/xmldom/xmldom/pull/514)/ [`#499`](https://github.com/xmldom/xmldom/issues/499)

Thank you, [@qtow](https://github.com/qtow), for your contributions

### Fixed

- Set nodeName property in ProcessingInstruction [`#509`](https://github.com/xmldom/xmldom/pull/509) / [`#505`](https://github.com/xmldom/xmldom/issues/505)

Thank you, [@cjbarth](https://github.com/cjbarth), for your contributions

### Fixed

- extend list of HTML entities [`#489`](https://github.com/xmldom/xmldom/pull/489)

Thank you, [@zorkow](https://github.com/zorkow), for your contributions

### Fixed

- properly parse closing where the last attribute has no value [`#485`](https://github.com/xmldom/xmldom/pull/485) / [`#486`](https://github.com/xmldom/xmldom/issues/486)

Thank you, [@bulandent](https://github.com/bulandent), for your contributions

### Fixed

- Properly check nodes before replacement [`#457`](https://github.com/xmldom/xmldom/pull/457) / [`#455`](https://github.com/xmldom/xmldom/issues/455) / [`#456`](https://github.com/xmldom/xmldom/issues/456)

Thank you, [@edemaine](https://github.com/edemaine), [@pedro-l9](https://github.com/pedro-l9), for your contributions

## 0.7.8

[Commits](https://github.com/xmldom/xmldom/compare/0.7.7...0.7.8)

### Fixed

- fix: Restore ES5 compatibility [`#452`](https://github.com/xmldom/xmldom/pull/452) / [`#453`](https://github.com/xmldom/xmldom/issues/453)

Thank you, [@fengxinming](https://github.com/fengxinming), for your contributions

## 0.7.7

[Commits](https://github.com/xmldom/xmldom/compare/0.7.6...0.7.7)

### Fixed

- Security: Prevent inserting DOM nodes when they are not well-formed [`CVE-2022-39353`](https://github.com/xmldom/xmldom/security/advisories/GHSA-crh6-fp67-6883)
  In case such a DOM would be created, the part that is not well-formed will be transformed into text nodes, in which xml specific characters like `<` and `>` are encoded accordingly.
  In the upcoming version 0.9.0 those text nodes will no longer be added and an error will be thrown instead.
  This change can break your code, if you relied on this behavior, e.g. multiple root elements in the past. We consider it more important to align with the specs that we want to be aligned with, considering the potential security issues that might derive from people not being aware of the difference in behavior.
  Related Spec: <https://dom.spec.whatwg.org/#concept-node-ensure-pre-insertion-validity>

Thank you, [@frumioj](https://github.com/frumioj), [@cjbarth](https://github.com/cjbarth), [@markgollnick](https://github.com/markgollnick) for your contributions

## 0.7.6

[Commits](https://github.com/xmldom/xmldom/compare/0.7.5...0.7.6)

### Fixed
- Avoid iterating over prototype properties [`#441`](https://github.com/xmldom/xmldom/pull/441) / [`#437`](https://github.com/xmldom/xmldom/pull/437) / [`#436`](https://github.com/xmldom/xmldom/issues/436)

Thank you, [@jftanner](https://github.com/jftanner), [@Supraja9726](https://github.com/Supraja9726) for your contributions

## 0.7.5

[Commits](https://github.com/xmldom/xmldom/compare/0.7.4...0.7.5)

### Fixes:

- Preserve default namespace when serializing [`#319`](https://github.com/xmldom/xmldom/issues/319) / [`#321`](https://github.com/xmldom/xmldom/pull/321)
  Thank you [@lupestro](https://github.com/lupestro)

## 0.7.4

[Commits](https://github.com/xmldom/xmldom/compare/0.7.3...0.7.4)

### Fixes:

- Restore ability to parse `__prototype__` attributes [`#315`](https://github.com/xmldom/xmldom/pull/315)
  Thank you [@dsimpsonOMF](https://github.com/dsimpsonOMF)

## 0.7.3

[Commits](https://github.com/xmldom/xmldom/compare/0.7.2...0.7.3)

### Fixes:

- Add doctype when parsing from string [`#277`](https://github.com/xmldom/xmldom/issues/277) / [`#301`](https://github.com/xmldom/xmldom/pull/301)
- Correct typo in error message [`#294`](https://github.com/xmldom/xmldom/pull/294)
  Thank you [@rrthomas](https://github.com/rrthomas)

### Refactor:

- Improve exports & require statements, new main package entry [`#233`](https://github.com/xmldom/xmldom/pull/233)

### Docs:

- Fix Stryker badge [`#298`](https://github.com/xmldom/xmldom/pull/298)
- Fix link to help-wanted issues [`#299`](https://github.com/xmldom/xmldom/pull/299)

### Chore:

- Execute stryker:dry-run on branches [`#302`](https://github.com/xmldom/xmldom/pull/302)
- Fix stryker config [`#300`](https://github.com/xmldom/xmldom/pull/300)
- Split test and lint scripts [`#297`](https://github.com/xmldom/xmldom/pull/297)
- Switch to stryker dashboard owned by org [`#292`](https://github.com/xmldom/xmldom/pull/292)

## 0.7.2

[Commits](https://github.com/xmldom/xmldom/compare/0.7.1...0.7.2)

### Fixes:

- Types: Add index.d.ts to packaged files [`#288`](https://github.com/xmldom/xmldom/pull/288)
  Thank you [@forty](https://github.com/forty)

## 0.7.1

[Commits](https://github.com/xmldom/xmldom/compare/0.7.0...0.7.1)

### Fixes:

- Types: Copy types from DefinitelyTyped [`#283`](https://github.com/xmldom/xmldom/pull/283)
  Thank you [@kachkaev](https://github.com/kachkaev)

### Chore:
- package.json: remove author, maintainers, etc. [`#279`](https://github.com/xmldom/xmldom/pull/279)

## 0.7.0 

[Commits](https://github.com/xmldom/xmldom/compare/0.6.0...0.7.0)

Due to [`#271`](https://github.com/xmldom/xmldom/issue/271) this version was published as
- unscoped `xmldom` package [to github only](https://github.com/xmldom/xmldom/releases/tag/0.7.0%2Bunscoped) (git tags [`0.7.0`](https://github.com/xmldom/xmldom/tree/0.7.0) and [`0.7.0+unscoped`](https://github.com/xmldom/xmldom/tree/0.7.0%2Bunscoped))
- scoped `@xmldom/xmldom` package to npm (github release/tag [`0.7.0+scoped`](https://github.com/xmldom/xmldom/releases/tag/0.7.0%2Bscoped))
For more details look at [`#278`](https://github.com/xmldom/xmldom/pull/278#issuecomment-902172483)

### Fixes:

- Security: Misinterpretation of malicious XML input [`CVE-2021-32796`](https://github.com/xmldom/xmldom/security/advisories/GHSA-5fg8-2547-mr8q)
- Implement `Document.getElementsByClassName` as specified [`#213`](https://github.com/xmldom/xmldom/pull/213), thank you [@ChALkeR](https://github.com/ChALkeR)
- Inherit namespace prefix from parent when required [`#268`](https://github.com/xmldom/xmldom/pull/268)
- Handle whitespace in closing tags [`#267`](https://github.com/xmldom/xmldom/pull/267)
- Update `DOMImplementation` according to recent specs [`#210`](https://github.com/xmldom/xmldom/pull/210)  
  BREAKING CHANGE: Only if you "passed features to be marked as available as a constructor arguments" and expected it to "magically work".
- No longer serializes any namespaces with an empty URI [`#244`](https://github.com/xmldom/xmldom/pull/244)   
  (related to [`#168`](https://github.com/xmldom/xmldom/pull/168) released in 0.6.0)  
  BREAKING CHANGE: Only if you rely on ["unsetting" a namespace prefix](https://github.com/xmldom/xmldom/pull/168#issuecomment-886984994) by setting it to an empty string 
- Set `localName` as part of `Document.createElement` [`#229`](https://github.com/xmldom/xmldom/pull/229), thank you [@rrthomas](https://github.com/rrthomas)

### CI

- We are now additionally running tests against node v16
- Stryker tests on the master branch now run against node v14

### Docs

- Describe relations with and between specs: [`#211`](https://github.com/xmldom/xmldom/pull/211), [`#247`](https://github.com/xmldom/xmldom/pull/247)

**Comment by karfau** [comment]
(The issue has been resolved, the package is now published as `@xmldom/xmldom`)

~FYI: [We are currently not able to publishing this package to npm](https://github.com/xmldom/xmldom/issues/271), you can subscribe to that issue to get notified about status updates.~

~As a workaround I uploaded the package in the [release on github](https://github.com/xmldom/xmldom/releases/tag/0.7.0).~

~It's also possible to [install it using the github tag](https://docs.npmjs.com/cli/v6/commands/npm-install), 
either by running `npm install github:xmldom/xmldom#0.7.0`~

~or by manually modifying your package.json:~
```diff
-  "xmldom": "0.6.0"
+  "xmldom": "github:xmldom/xmldom#0.7.0"
```

Just keep in mind: In case you are using a tool to provide PRs for package updates, that you might need change it back to just the version string, before you might get future updates. (Check the details of your tool.)

**Reply by RopoMen** [reply]
@karfau what is the current ETA for the npm publish?

**Reply by karfau** [reply]
We don't have any ETA, since we did not receive any information from npm yet.

We will publish it as soon as the access has been restored.

Last checked 2021-08-09T16:35:00 UTC (#271 was updated)

**Reply by RopoMen** [reply]
NPM Security Advisory published yesterday https://www.npmjs.com/advisories/1769

**Reply by karfau** [reply]
> Reported by Anonymous

Not sure what that means. But the process of "grabbing" the information from the github advisory already took a  while. (Not sure if we would have been able to speed it up or if that is automated.)
I'm trying on a daily basis if we are able to publish the package, but I assume the security advisory is independent from the ownership issue.

**Reply by jquintozamora** [reply]
Thank you for looking at this @karfau , 
I'm looking forward to have the new version in npm as in my company we are not allowed to import directly from github ( by policy ).

**Reply by forty** [reply]
@karfau is there anything we can do to help this going faster? How was the access lost? I'm happy to tweet @ npm or send emails if that can help ;)

**Reply by pkalantri-ebsco** [reply]
Any update on this? We are having the same issue at my clients as @jquintozamora where we can't import this directly via github (by policy).

**Reply by brody2consult** [reply]
Please watch #271 for the updates, thanks.

We may need 1-2 weeks or so to discuss how we lost access, etc.

**Reply by Schwad** [reply]
👀 

Watching; thanks for the updates

**Reply by slawekjaranowski** [reply]
after 

```
npm install github:xmldom/xmldom#0.7.0
```

and commit I have error on CI..

```
npm ERR! Error while executing:
npm ERR! /usr/bin/git ls-remote -h -t ssh://git@github.com/xmldom/xmldom.git
npm ERR! 
npm ERR! Warning: Permanently added the RSA host key for IP address '192.30.255.112' to the list of known hosts.
npm ERR! git@github.com: Permission denied (publickey).
npm ERR! fatal: Could not read from remote repository.
``` 

https://github.com/s4u/maven-settings-action/pull/133

**Comment by XhmikosR** [comment]
@karfau can't you just cut a 0.7.1?

BTW please try to keep security fixes as patch versions because right now every downstream package has to manually update to the new version :/

**Reply by djejaquino** [reply]
@XhmikosR If you are using yarn, you can use [selective dependency resolutions](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/) and point to GitHub's registry.

`package.json`
```jsonc
{
  ...
  "resolutions": {
    // globally
    "xmldom": "github:xmldom/xmldom#0.7.0",
    // targeted
    "downstream-dependency/xmldom": "github:xmldom/xmldom#0.7.0",
  }
...
}
```

**Reply by XhmikosR** [reply]
I don't use yarn nor I want to use github directly. It's not safe to use packages through GitHub.

**Reply by djejaquino** [reply]
`xmldom` is a zero-dependency package, which means installing it from GitHub is as safe as installing from NPM.

Anyways, I guess you will have to wait  🤷

**Reply by karfau** [reply]
@XhmikosR 
> can't you just cut a 0.7.1?

To solve what problem? Do you mean 0.6.1?

I'm aware that we are currently releasing security releases together with breaking changes. (As the version of xmldom indicates, it is not stable.)

The "breaking" changes released in 0.7.0 hopefully do not really affect anybody that's why we did not release a version 0.6.1, but it would still be possible if somebody is not able to upgrade to 0.7.0.

I'm thinking about adding semantic releases in the near future, so any change would be released right away. But the priority/resources for this are not clear yet

**Reply by karfau** [reply]
@XhmikosR see https://github.com/xmldom/xmldom/discussions/270#discussioncomment-1140323

**Reply by admosity** [reply]
![image](https://user-images.githubusercontent.com/4655972/128587177-6091629f-d7fd-4f47-ad19-abbe34ed0c88.png)

![image](https://user-images.githubusercontent.com/4655972/128587200-cbb1ae11-eeb6-41b6-ac19-5099365b301a.png)


Works with https://www.npmjs.com/package/npm-force-resolutions for npm users.

**Comment by hansemannn** [comment]
Thanks for being so responsive on this, @karfau!

**Comment by karfau** [comment]
Please comment on this thread with some details why you can not upgrade to 0.7.0 and need a 0.6.1 for the security fix.
**This is not about the issue regarding publishing to npm!**

**Reply by forty** [reply]
Some packages depends on ^0.6.0. publishing a 0.6.1 would allow people who depends on those packages to get the new release without waiting for those packages maintainers to do the update. This would still require publishing to npm though, so I think it's a minor issue for now.

**Reply by karfau** [reply]
@forty This makes sense.
As soon as we are able to publish again, I will release this one as well.

**Reply by forty** [reply]
Thanks!

**Reply by karfau** [reply]
According to the download stats in the npm registry it even makes sense to publish 0.5.1:
![image](https://user-images.githubusercontent.com/135657/128591622-8f6def0d-2ba3-4bf1-bcb3-4889f8c99d7b.png)

I prepared the branches for the two versions already.

**Reply by karfau** [reply]
@forty due to the new package name I would expect that people need to anyways do a manual step to update.

Do you think we should still invest the time to publish 0.6.1 (or even 0.5.1) with the new package name? And for what reason?

**Reply by forty** [reply]
Agreed, the package name change makes it useless to publish updates on older versions 👍

**Comment by andurmon** [comment]
Hi @karfau any updates on this issue? the temporary fix of using "xmldom": "github:xmldom/xmldom#0.7.0" in the package-lock works smoothly locally. However my team has issues deploying and installing packages that are not in the npm registry. Apparently our deploy pipeline simply cannot find anything outside the npm registry.

**Reply by karfau** [reply]
According to the current status of #271 we will need to publish the package under the `@xmldom` scope, we "just" to decide on the name.

**Reply by forty** [reply]
@xmldom/xmldom ?

**Reply by RopoMen** [reply]
@andurmon if you are using `nmp ci / i´ on your CI pipeline, remember that your CI runner or container that is executing that command must have `git` installed. If you are using yarn, then you do not need ´git´ as your dependency. 

Whyt it works locally? well, most developers has git already installed.

**Reply by karfau** [reply]
We will soon publish the fixed versions under the `@xmldom` scope to npm, the "fully qualified name" of the library will switch to `@xmldom/xmldom`. For details see #278

**Reply by karfau** [reply]
0.7.0 has been released yesterday.
It had a regression regarding types, since the unscoped package was typed via `@types/xmldom`.
0.7.1 has been released just some minutes ago and includes those types.
